algopy_testing/src/algopy/op.py

```diff
@@ -0,0 +1,159 @@
+import hashlib
+from enum import Enum
+
+import algosdk
+import coincurve
+import nacl.exceptions
+import nacl.signing
+from algopy_testing.constants import MAX_BYTES_SIZE
+from algopy_testing.contexts import state_context
+from Cryptodome.Hash import keccak
+from ecdsa import (  # type: ignore  # noqa: PGH003
+    BadSignatureError,
+    NIST256p,
+    SECP256k1,
+    VerifyingKey,
+)
+
+from algopy import Bytes, UInt64
+from algopy.utils import as_bytes
+
+
+class ECDSA(Enum):
+    Secp256k1 = 0
+    Secp256r1 = 1
+
+
+def sha256(a: Bytes | bytes, /) -> Bytes:
+    input_value = as_bytes(a, max_size=MAX_BYTES_SIZE)
+    return Bytes(hashlib.sha256(input_value).digest())
+
+
+def sha3_256(a: Bytes | bytes, /) -> Bytes:
+    input_value = as_bytes(a, max_size=MAX_BYTES_SIZE)
+    return Bytes(hashlib.sha3_256(input_value).digest())
+
+
+def keccak256(a: Bytes | bytes, /) -> Bytes:
+    input_value = as_bytes(a, max_size=MAX_BYTES_SIZE)
+    hashed_value = keccak.new(data=input_value, digest_bits=256)
+    return Bytes(hashed_value.digest())
+
+
+def sha512_256(a: Bytes | bytes, /) -> Bytes:
+    input_value = as_bytes(a, max_size=MAX_BYTES_SIZE)
+    hash_object = hashlib.new("sha512_256", input_value)
+    return Bytes(hash_object.digest())
+
+
+def ed25519verify_bare(a: Bytes | bytes, b: Bytes | bytes, c: Bytes | bytes, /) -> bool:
+    a, b, c = (as_bytes(x, max_size=MAX_BYTES_SIZE) for x in [a, b, c])
+
+    try:
+        verify_key = nacl.signing.VerifyKey(c)
+        verify_key.verify(a, b)
+    except nacl.exceptions.BadSignatureError:
+        return False
+    else:
+        return True
+
+
+def ed25519verify(a: Bytes | bytes, b: Bytes | bytes, c: Bytes | bytes, /) -> bool:
+    """
+    NOTE: Differs from AVM behavior since it does not having access to the precompiled
+    contract hash.
+    Currently works by forwarding the call to ed25519verify_bare
+    """
+    with state_context() as state:
```

Comment:
If we use contextvar (see other comment) then this op should just try and access the current context with the error below if no current context is set

The testing code should be responsible for actually setting up the context using a `with` block

Reply:
@daniel-makerx let me know if the refinement makes sense, i do check on any context first but then it also checks on the specific program_hash field (which we set via context dictionary that user passes when state_context generator is invoked)